### DISCUSSION TITLE
Potentially a bug creating .tlb files for early binding Com Server?

### DISCUSSION THREAD

**Discussion by cirrusone** [Excel-DNA/ExcelDna#729] [Q&A] [answered]
I need to create a COM server that can raise events in VBA (using WithEvents) so this requires early binding and adding reference to VBA Tools/References.

Essentially I followed the [ComServer](https://github.com/Excel-DNA/Samples/tree/master/ComServer) example but have tried to adapt it to `Excel-DNA 1.8.0` and `SDK-style project files`

The project works fine late bound using `Dim lib As Object : Set lib = CreateObject("TestProject.ComAddin")` but without events. Changing the project file to the following 

```
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<TargetFramework>net8.0-windows</TargetFramework>
		<UseWindowsForms>true</UseWindowsForms>
		<ImplicitUsings>disable</ImplicitUsings>
		<Nullable>disable</Nullable>
	</PropertyGroup>

	<PropertyGroup>
		<ExcelAddInTlbExp>"C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8 Tools\x64\TlbExp.exe"</ExcelAddInTlbExp>
		<ExcelAddInTlbCreate>true</ExcelAddInTlbCreate>
		<ExcelAddInComServer>true</ExcelAddInComServer>
	</PropertyGroup>

	<ItemGroup>
	  <PackageReference Include="Excel-DNA" Version="1.8.0" />
	  <PackageReference Include="ExcelDna.Integration" Version="1.8.0" />
	  <PackageReference Include="Microsoft.Web.WebView2" Version="1.0.2792.45" />
	  <PackageReference Include="Serilog" Version="4.0.2" />
	  <PackageReference Include="Serilog.Expressions" Version="5.0.0" />
	  <PackageReference Include="Serilog.Sinks.File" Version="6.0.0" />
	</ItemGroup>

</Project>
```

Does not generate the .tlb file and gives the following error on build (the path to TlbExp.exe is correct on my machine).

```
1>------ Rebuild All started: Project: TestProject, Configuration: Debug Any CPU ------
Restored C:\Users\o\Desktop\TestProject\TestProject\TestProject.csproj (in 9 ms).
1>---
1>---
1>ExcelDnaSetLaunchSettings: EXCEL.EXE path for debugging: C:\Program Files\Microsoft Office\root\Office16\EXCEL.EXE
1>ExcelDnaSetLaunchSettings: Add-In for debugging: bin\Debug\net8.0-windows\TestProject-AddIn64.xll
1>TestProject -> C:\Users\o\Desktop\TestProject\TestProject\bin\Debug\net8.0-windows\TestProject.dll
1>---
1>MSBUILD : ExcelDnaBuild error DNA126955310: The system cannot find the file specified
1>MSBUILD : ExcelDnaBuild error DNA126955310: System.ComponentModel.Win32Exception (0x80004005): The system cannot find the file specified
1>MSBUILD : ExcelDnaBuild error DNA126955310:    at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
1>MSBUILD : ExcelDnaBuild error DNA126955310:    at ExcelDna.AddIn.Tasks.Utils.ProcessRunner.Run(String fileName, String arguments, String appName, IBuildLogger log)
1>MSBUILD : ExcelDnaBuild error DNA126955310:    at ExcelDna.AddIn.Tasks.CreateExcelAddIn.TryCreateTlb()
1>MSBUILD : ExcelDnaBuild error DNA126955310:    at ExcelDna.AddIn.Tasks.CreateExcelAddIn.Execute()
1>Done building project "TestProject.csproj" -- FAILED.
```

I've tried to simplify things by starting with the sample [ComServer](https://github.com/Excel-DNA/Samples/tree/master/ComServer) project but still run into problems with no .tlb file being generated. The only line I've changed in the project is removing `net472` from the proj file:

`TargetFrameworks>net6.0-windows</TargetFrameworks>`

In VBA I get `MISSING: ComServer` as there is no .tlb file generated.

![Untitled](https://github.com/user-attachments/assets/fb23f7a5-a069-4652-bfdf-8d0de515235d)

Is this a bug on net8.0 / net6.0 or am I missing something obvious?  I can post more if required.

**Comment by govert** [comment, marked as answer]
I'm sorry the samples and documentation for this are such a mess.

Can you try to make the project like this, with `<ExcelAddInTlbCreate>false</ExcelAddInTlbCreate>` or removed, and adding the `DsCom` package reference explicitly (`dSPACE.Runtime.InteropServices.BuildTasks`)?

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0-windows</TargetFramework>
    <UseWindowsForms>true</UseWindowsForms>
    <ImplicitUsings>disable</ImplicitUsings>
    <Nullable>disable</Nullable>
  </PropertyGroup>

  <PropertyGroup>
    <ExcelAddInTlbExp>C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8 Tools\x64\TlbExp.exe</ExcelAddInTlbExp>
    <ExcelAddInTlbCreate>false</ExcelAddInTlbCreate>
    <ExcelAddInComServer>true</ExcelAddInComServer>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="ExcelDna.AddIn" Version="1.8.0" />
    <PackageReference Include="dSPACE.Runtime.InteropServices.BuildTasks" Version="1.11.0" />
  </ItemGroup>

</Project>
```

**Comment by cirrusone** [comment]
Excellent, it now works after adding DsCom. Thank you.

On a minor note in case anyone else stumbles on this; adding the nuget packages in following order works fine

```
<PackageReference Include="Excel-DNA" Version="1.8.0" />
<PackageReference Include="dSPACE.Runtime.InteropServices.BuildTasks" Version="1.11.0" />
<PackageReference Include="ExcelDna.Integration" Version="1.8.0" />
```

but adding  nuget packages in following order did not

```
<PackageReference Include="Excel-DNA" Version="1.8.0" />
<PackageReference Include="ExcelDna.Integration" Version="1.8.0" />
<PackageReference Include="dSPACE.Runtime.InteropServices.BuildTasks" Version="1.11.0" />
```

raising following errors. It might be due to requiring VS restart though

```
Severity	Code	Description	Project	File	Line	Suppression State
Error		Package 'dscom 1.11.0' has a package type 'DotnetTool' that is not supported by project 'TestProject'.				
Error	NU1212	Invalid project-package combination for dscom 1.11.0. DotnetToolReference project style can only contain references of the DotnetTool type 	TestProject	C:\Users\o\Desktop\TestProject\TestProject\TestProject.csproj	1	
Error	NU1202	Package dscom 1.11.0 is not compatible with net8.0-windows7.0 (.NETCoreApp,Version=v8.0). Package dscom 1.11.0 supports:
  - net6.0 (.NETCoreApp,Version=v6.0) / any
  - net8.0 (.NETCoreApp,Version=v8.0) / any	TestProject	C:\Users\o\Desktop\TestProject\TestProject\TestProject.csproj	1
```

**Comment by govert** [comment]
You don't need to add the package reference for `ExcelDna.Integration` and you can use `ExcelDna.AddIn` instead of `Excel-DNA`.

**Comment by cirrusone** [comment]
Thanks.

The .tlb is now being generated and I can add the reference but I still have problems getting the basic COM server working. Calling a method in the class library gives a VBA error

`Compile error: Object library feature not supported`

I'm just using the sample ComServer project at the moment with no changes. If I remove the following line then the class library will then call `ml.Fire` but add it back and compile error again.

`Public WithEvents mh As MessageHandler`

Any idea what is causing this? 

![Untitled](https://github.com/user-attachments/assets/046ed0a2-99aa-4751-aa08-59d8288cf190)
![Untitled2](https://github.com/user-attachments/assets/3a07d98a-3cd4-40bf-aa59-4e6e3cba7805)

**Comment by govert** [comment]
I had this code

```c#
using System.Diagnostics;
using System.Runtime.InteropServices;

namespace TestComServer18
{
    // Step 1: Defines an event sink interface (MessageEvents) to be implemented by the COM sink.
    [ComVisible(true)]
    [InterfaceType(ComInterfaceType.InterfaceIsIDispatch)]
    public interface MessageEvents
    {
        void NewMessage(string s);
    }

    // Step 2: Defines a custom interface for the MessageHandler class.
    [ComVisible(true)]
    [InterfaceType(ComInterfaceType.InterfaceIsIDispatch)]
    public interface IMessageHandler
    {
        void FireNewMessageEvent(string s);
    }

    // Step 3: Connects the event sink interface to a class by passing the namespace and event sink interface
    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.None)]  // Avoid using AutoDual
    [ComSourceInterfaces(typeof(MessageEvents))]
    public class MessageHandler : IMessageHandler
    {
        public delegate void NewMessageDelegate(string s);
        public event NewMessageDelegate NewMessage;
        public MessageHandler() { }

        public void FireNewMessageEvent(string s)
        {
            Debug.Print($"New Message {s}");
            if (NewMessage != null)
            {
                Debug.Print($"Invoke {s}");
                NewMessage.Invoke(s);
            }
        }
    }
}

```

**Comment by govert** [comment]
The `AutoDual` interface is not supported by DsCom.

**Comment by govert** [comment]
When I had the AutoDual, DsCom failed at build time though.

**Comment by cirrusone** [comment]
Sorry, still can't get it working. Step2 was not in original sample project so have added.

Also added  `[ComDefaultInterface(typeof(IMessageHandler))]`. 

```
    [ComVisible(true)]
    [ClassInterface(ClassInterfaceType.None)]  // Avoid using AutoDual
    [ComSourceInterfaces(typeof(MessageEvents))]
    [ComDefaultInterface(typeof(IMessageHandler))]
    public class MessageHandler : IMessageHandler
    ....
```

The VBA project compiles after adding new tlb file but now get 'Class not registered' error.

![Untitled](https://github.com/user-attachments/assets/66d94d33-301d-4d66-b0fa-c3ccb24c0007)

Sorry for taking up so much of your time. If I can get a simple working ComServer project I'll be okay afterwards.

**Comment by govert** [comment]
Be careful with VBA projects that might cache the library. Then further changes are not reflected.

You can try
https://github.com/govert/TestComServer18

You need to build and load the add-in, then open the TestWorkbook.xlsm, Alt+F11, then run the `RunMessageHandlerTest` in the module.

I see 
![image](https://github.com/user-attachments/assets/a25ffa0d-2800-4e8a-86b7-2a8dfb1d7ac0)

**Comment by cirrusone** [comment]
TestComServer18 works perfectly, although not quite sure what difference is between earlier examples.

Is there a way to ensure VBA clears cached references rather than changing project name or rolling back to an earlier Windows image?
